DEVELOPMENT.md

```diff
@@ -0,0 +1,131 @@
+# Design of structs
+
+Example 'literal' of a struct (the explicit type is optional):
+
+    let s : {a : Int, b : Bool, c : {x : Int, y : Int}} = {a : Int = 2, b = false && true, c : {x : Int, y : Int} = {x = 3, y = (let z = 3+4; z)}};
+
+The type of struct is determined by the names and types of its fields (in the unordered manner). Field names have to be unique. Scopes of init expressions are independent.
+
+Type is a subtype of Type B iff for every field of A with type T_A and name N, there is a field of B with the same name N and type T_B such that T_A is a subtype of T_B.
+
+For example, the type `{x: Int, y : {a : Int}}` is a subtype of `{x : Int, y: {a : Int, b : Bool}, z : Bool}`. There is no relation between types `Int` and `{a: Int}`.
+
+Richer type hierarchy implies that we need to implement some logic for the hierarchy of function types.
+
+## Implementation details
+
+Disclaimer: The usage of the String type in the following code snippets is up for discussion and change to the more appropriate type.
+
+Consider the following snippet:
+
+    let s = {x = 5, y = {a = 3, b = true}};
+    let f = [] -> Int => s.x + s.y.a;
+
+In theory, we do not need to keep the whole `s` in memory. We could place `s.x` and `s.y.a` on the stack for the static link to work inside `f`, but `x.y.b` could be placed in a virtual register.
+
+To achieve this effect, we propose the following interface for variables:
+
+```kotlin
+sealed class VariableInfo
+class PrimitiveVariable : VariableInfo()
+class StructVariable(val fields: Map<String, VariableInfo>) : VariableInfo()
+
+sealed class Variable(open val info: VariableInfo)
+
+class SourceVariable(val definition: Definition, override val info : VariableInfo) : Variable(info)
+
+sealed class AuxVariable(info: VariableInfo) : Variable(info)
+class SpillVariable(override val info: PrimitiveVariable) : AuxVariable(info)
+class StaticLinkVariable(override val info: PrimitiveVariable) : AuxVariable(info)
+
+class ReturnVariable(override val info: VariableInfo) : Variable(info)
+```
+
+The previous snippet should introduce the following `VariableInfo`:
+
+    sx = PrimitiveVariable()
+    sya = PrimitiveVariable()
+    syb = PrimitiveVariable()
+    sy = StructVariable({"a" : sya, "b" : syb})
+    s = StructVariable({"x" : sx, "y" : sy})
+
+### AST
+
+Introducing the following node type should be enough (we assume that its semantics are clear).
+
+```kotlin
+class SubfieldAccess(val expr: Expression, val field: String)
+```
+
+TypeChecker should check if every subfield access is valid.
+
+### Variable Analysis / Name Resolution
+
+New logic to determine all source variables and their usage is needed. FunctionHandler should receive analyzed `SourceVariable` corresponding to every subfield, and not the `Definition` like now.
```

Comment:
What if we get rid of types of Variable? It is only used in `generateVariableAccess` and even there we do not really make use of it - it could be replaced with a map from Variable to function handler in which it was defined